Objects/moduleobject.c

```diff
@@ -348,6 +348,69 @@ PyModule_FromDefAndSpec2(struct PyModuleDef* def, PyObject *spec, int module_api
 }
 
 int
+PyModule_ExecInModule(PyObject *module, PyModuleDef *def)
+{
+    PyModuleDef_Slot *cur_slot;
+    const char *name;
+    PyObject *nameobj;
+
+    if (!PyModule_Check(module)) {
+        PyErr_BadInternalCall();
+        goto error;
+    }
+
+    nameobj = PyModule_GetNameObject(module);
+    if (nameobj == NULL) {
+        goto error;
+    }
+
+    name = PyUnicode_AsUTF8(nameobj);
+    if (name == NULL) {
+        goto error;
+    }
+
+
+
+    if (((PyModuleObject *)module)->md_state != NULL) {
+        PyErr_Format(
+            PyExc_ImportError,
+            "Module already initialized.",
+            name);
+        goto error;
+    }
+
+    for (cur_slot = def->m_slots; cur_slot && cur_slot->slot; cur_slot++) {
+        if (cur_slot->slot == Py_mod_create) {
+            /* Modules with Py_mod_create cannot be directly executed */
```

Comment:
The interpreter itself creates `__main__` during startup, so it needs to be able to hand the loader that pre-existing module rather than asking the asking the loader to create a new one. If we don't do that, then we end up with subtle problems like `-i` and `PYTHONINTERACTIVE=1` not working as expected.

That does raise an interesting difference with `runpy.run_module` though: that *can* ask the loader to create its own module object, so it should be able to handle arbitrary extension modules, even those that define `Py_mod_create`.

If that poses a potential problem for Cython though, then it would make sense to look more closely at the feasibility of just lifting the restriction against replacing `__main__`, and instead improve the post-mortem introspection code to better handle the case where `sys.modules["__main__"]` isn't the main module the interpreter originally created.